contrib/direnv

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env bash
+
+# Look for the .config folder in the current directory and up
+find_prj_config() (
+	local old_pwd
+	while [[ $old_pwd != $PWD ]]; do
+		if [[ -d .config ]]; then
+			echo "$PWD"
+			return 0
+		fi
+		old_pwd=$PWD
+		cd ..
+	done
+	# We're at the top and didn't find anything
+	log_error "ERROR: could not find project root" >&2
+	return 1
+)
+
+{
+	: "${XDG_CONFIG_HOME:=${HOME}/.config}"
+	: "${XDG_CACHE_HOME:=${HOME}/.cache}"
+	: "${XDG_DATA_HOME:=${HOME}/.local/share}"
+	: "${XDG_STATE_HOME:=${HOME}/.local/state}"
+	: "${XDG_RUNTIME_DIR:=${HOME}/.local/state}"
+
+	: "${GIT_DIR:=$(git rev-parse --absolute-git-dir 2>/dev/null)}"
+
+	# PRJ_ROOT
+	_prj_root="${GIT_DIR%/.git}"
+	: "${PRJ_ROOT:=${_prj_root:=$(find_prj_config)}}"
+
+	# PRJ_CONFIG_HOME - always local to the project
+	: "${PRJ_CONFIG_HOME:=${PRJ_ROOT}/.config}"
+	mkdir -p "${PRJ_CONFIG_HOME}"
+
+	if [[ -z "${PRJ_ID:-}" && -f "${PRJ_CONFIG_HOME}/prj_id" ]]; then
+		PRJ_ID=$(<"${PRJ_CONFIG_HOME}/prj_id")
+	fi
+
+	# PRJ_RUNTIME_DIR - always local to the project
+	: "${PRJ_RUNTIME_DIR:=${PRJ_ROOT}/.run}"
+	mkdir -p "${PRJ_RUNTIME_DIR}"
+
+	# PRJ_CACHE_HOME - shared if PRJ_ID is set
+	if [[ -z "${PRJ_CACHE_HOME:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_CACHE_HOME="${XDG_CACHE_HOME}/prj/${PRJ_ID}"
+		else
+			PRJ_CACHE_HOME="${PRJ_ROOT}/.cache"
+		fi
+	fi
+	mkdir -p "${PRJ_CACHE_HOME}"
+
+	# PRJ_DATA_HOME - shared if PRJ_ID is set
+	if [[ -z "${PRJ_DATA_HOME:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_DATA_HOME="${XDG_DATA_HOME}/prj/${PRJ_ID}"
+		else
+			PRJ_DATA_HOME="${PRJ_ROOT}/.local/share"
+		fi
+	fi
+	mkdir -p "${PRJ_DATA_HOME}"
+
+	# PRJ_STATE_HOME - shared if PRJ_ID is set
+	if [[ -z "${PRJ_STATE_HOME:-}" ]]; then
+		if [[ -n "${PRJ_ID:-}" ]]; then
+			PRJ_STATE_HOME="${XDG_STATE_HOME}/prj/${PRJ_ID}"
+		else
+			PRJ_STATE_HOME="${PRJ_ROOT}/.local/state"
+		fi
```

Comment:
Instead, we could rename `.local/share` to `${PRJ_ROOT}/.data`. I'm not sure why the XDG spec picked `.local/share`, maybe because of NFS mounts?

Reply:
Agreed! That makes actually more intuitive sense!